### PR TITLE
nanopb: init at 0.4.1

### PR DESCRIPTION
**Pull request by kalbasit**
###### Motivation for this change

Nanopb [0.4.1](https://github.com/nanopb/nanopb/releases/tag/0.4.1) is a small code-size Protocol Buffers implementation in ansi C. It is especially suitable for use in microcontrollers, but fits any memory restricted system.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
interesting way to create tests... not sure if i love or hate it :)

**Comment by kalbasit** [comment]
> interesting way to create tests... not sure if i love or hate it :)

Would you like to discuss your concerns? It'd be great to get this merged soon.

File: pkgs/development/libraries/nanopb/default.nix [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      --subst-var-by python ${pythonRuntime}/bin/python \
+      --subst-var-by out $out
+    chmod +x $out/bin/protoc-gen-nanopb
+  '';
+
+  passthru.tests = {
+    simple-proto2 = callPackage ./test-simple-proto2 {};
+    simple-proto3 = callPackage ./test-simple-proto3 {};
+    message-with-annotations = callPackage ./test-message-with-annotations {};
+    message-with-options = callPackage ./test-message-with-options {};
```
Comment (Mic92) [inline pkgs/development/libraries/nanopb/default.nix:59]: I think @Profpatsch had a similar idea once.

**Comment by Mic92** [comment]
> > interesting way to create tests... not sure if i love or hate it :)
> 
> Would you like to discuss your concerns? It'd be great to get this merged soon.

I would split-off the test part of this PR. Using `passthru.tests` is an interesting idea. However I would like to discuss if we should separate nixosTests from other test derivations.

**Comment by kalbasit** [comment]
> > > interesting way to create tests... not sure if i love or hate it :)
> > 
> > 
> > Would you like to discuss your concerns? It'd be great to get this merged soon.
> 
> I would split-off the test part of this PR. Using `passthru.tests` is an interesting idea. However I would like to discuss if we should separate nixosTests from other test derivations. If you break change into two commits, I would cherry-pick one of them two masters.

This isn't the first package to do so:
- [Bazel](https://github.com/NixOS/nixpkgs/blob/19aac2413ae2810a47850f165a592cbe0d06f744/pkgs/development/tools/build-managers/bazel/bazel_3/default.nix#L192)
- [Phoronix test suite](https://github.com/NixOS/nixpkgs/blob/19aac2413ae2810a47850f165a592cbe0d06f744/pkgs/tools/misc/phoronix-test-suite/default.nix#L22-L24)
- [Jasmin](https://github.com/NixOS/nixpkgs/blob/19aac2413ae2810a47850f165a592cbe0d06f744/pkgs/development/compilers/jasmin/default.nix#L31-L33) <-- This is what I replicated in this PR.

Among others in `git grep passthru.tests -- pkgs/**/default.nix`

Also, ofborg has native support for `{pkg}.tests` now and it always tries to build `{pkg}` and `{pkg}.tests` see the [check for this PR](https://github.com/NixOS/nixpkgs/pull/89070/checks?check_run_id=715343347).

**Comment by kalbasit** [comment]
@Mic92 I split the commits as suggested. PTAL.

**Comment by jonringer** [comment]
@kalbasit sorry for not responding earlier.

The love part is having additional test which validate a package, the hate part is the amount of nix scaffolding needed to make it happen.

**Comment by kalbasit** [comment]
> @kalbasit sorry for not responding earlier.

No worries! Things can get busy at times, and it's completely understandable!

> The love part is having additional test which validate a package, the hate part is the amount of nix scaffolding needed to make it happen.

I get your point. I've had a bad experience with `protoc` and its plugins in the past. I want to make sure that updating the package does not just rely on the build passing but actually asserts that it works correctly. Nanopb itself does provide [tests](https://github.com/nanopb/nanopb/tree/master/tests), I'll explore that and see if I can swap out our tests in favor of those.

**Pull request merged**